base/unicode/utf8.jl

```diff
@@ -312,6 +312,11 @@ function convert(::Type{UTF8String}, a::Vector{UInt8}, invalids_as::AbstractStri
 end
 convert(::Type{UTF8String}, s::AbstractString) = utf8(bytestring(s))
 
+function convert{T<:Union{UInt16, UInt32}}(::Type{UTF8String}, a::Vector{T}, len=length(a)-1)
```

Comment:
`utf16(a::Vector{UInt16})` does *not* assume that `a` is NUL-terminated (if `a` ends with a NUL codeunit, then the resulting string contains a NUL).   So, it would be more consistent to use `len=length(a)` here, even if that means you need to do `UTF8String(buf, length(buf)-1)` at most of the call sites.